### ISSUE TITLE
advertise client URLs are resolved causing SSL verification complications

### ISSUE THREAD

**Issue by ssabo** [coreos/etcd#2488]
The advertised client URLs returned from /v2/members and /v2/machines are resolved to IP addresses. 

For clients that use these to discover the endpoint to query, this forces one of two conditions:
1. the SSL client must not verify the hostname it is connecting to against the names in the SSL cert
2. the server must have it's IP address in the CN or SAN of the SSL cert

Is there any reason we want to resolve the advertise client URLs or would it make sense to leave them unresolved?

The specific problem I am running into is with etcdctl. Lets say that my nodes are called:
https://node0.example.com:4001
https://node1.example.com:4001
https://node2.example.com:4001

And each one is configured with the server certificate that matches the domain "*.example.com"

Etcdctl will make a request against /v2/machines and will get back:
https://10.0.0.1:4001, https://10.0.0.2:4001, https://10.0.0.3:4001

Since the server's certificate does not have an IP address in its CN or SAN, etcdctl will fail to validate the server certificate when it makes the request for /v2/keys/

**Comment by marvell** [comment]
I have the same problem.

Start server:
```
netutil: Resolving etcd1.example.com:2380 to 10.0.0.11:2380
netutil: Resolving etcd1.example.com:2379 to 10.0.0.11:2379
etcd: listening for peers on https://0.0.0.0:2380
etcd: clientTLS: cert = /etc/etcd/certs/server.crt, key = /etc/etcd/certs/server.key, ca =
etcd: listening for client requests on https://0.0.0.0:2379
netutil: Resolving etcd1.example.com:2380 to 10.0.0.11:2380
etcdserver: name = etcd1
etcdserver: data dir = /var/lib/etcd
etcdserver: member dir = /var/lib/etcd/member
etcdserver: heartbeat = 100ms
etcdserver: election = 1000ms
etcdserver: snapshot count = 10000
etcdserver: advertise client URLs = https://10.0.0.11:2379
etcdserver: initial advertise peer URLs = https://10.0.0.11:2380
etcdserver: initial cluster = etcd1=https://etcd1.example.com:2380
etcdserver: start member 163b9e2c155990b9 in cluster d14e42a7105c79a9
raft: 163b9e2c155990b9 became follower at term 0
raft: newRaft 163b9e2c155990b9 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
raft: 163b9e2c155990b9 became follower at term 1
etcdserver: added local member 163b9e2c155990b9 [https://etcd1.example.com:2380] to cluster d14e42a7105c79a9
raft: 163b9e2c155990b9 is starting a new election at term 1
raft: 163b9e2c155990b9 became candidate at term 2
raft: 163b9e2c155990b9 received vote from 163b9e2c155990b9 at term 2
raft: 163b9e2c155990b9 became leader at term 2
raft.node: 163b9e2c155990b9 elected leader 163b9e2c155990b9 at term 2
etcdserver: published {Name:etcd1 ClientURLs:[https://10.0.0.11:2379]} to cluster d14e42a7105c79a9
```

Try to connect via curl - OK:
```
curl --cacert /etc/etcd/certs/ca.crt https://etcd1.example.com:2379/v2/keys -v
* Hostname was NOT found in DNS cache
*   Trying 10.0.0.11...
* Connected to etcd1.example.com (10.0.0.11) port 2379 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /etc/etcd/certs/ca.crt
  CApath: none
* SSL connection using TLS_RSA_WITH_AES_128_CBC_SHA
* Server certificate:
*   subject: CN=*.example.com,OU=EssentialSSL Wildcard,OU=Domain Control Validated
*   start date: фев 04 00:00:00 2015 GMT
*   expire date: фев 04 23:59:59 2016 GMT
*   common name: *.example.com
*   issuer: CN=COMODO RSA Domain Validation Secure Server CA,O=COMODO CA Limited,L=Salford,ST=Greater Manchester,C=GB
> GET /v2/keys HTTP/1.1
> User-Agent: curl/7.37.0
> Host: etcd1.example.com:2379
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Type: application/json
< X-Etcd-Cluster-Id: d14e42a7105c79a9
< X-Etcd-Index: 2
< X-Raft-Index: 30
< X-Raft-Term: 2
< Date: Thu, 12 Mar 2015 11:07:43 GMT
< Content-Length: 37
<
{"action":"get","node":{"dir":true}}
* Connection #0 to host etcd1.example.com left intact
```

And etcdctl:

```
etcdctl --ca-file=/etc/etcd/certs/ca.crt --peers="https://etcd1.example.com:2379" ls /
Error:  501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]

etcdctl --ca-file=/etc/etcd/certs/ca.crt --peers="https://etcd1.example.com:2379" --no-sync ls /
```

**Comment by bmurphy1976** [comment]
I think I am running into this problem.  Same cause, but different symptoms.

In my case, I have CNAME aliases for our nodes that resolve the private IP on the LAN and the public IP on the WAN.

We're using SSL to encrypt the connection and allow clients to connect from separate networks, however, because the servers advertise the private ip addresses instead of the dns name:

```
# curl -s http://127.0.0.1:4002/v2/machines
https://10.137.65.102:2379, https://10.147.206.41:2379, https://10.28.5.39:2379
```

clients that are on the WAN cannot connect to the private IP address and fail.

I'm passing:

```
--advertise-client-urls https://dns_cname_for_each_node:2379
```

at startup to each ETCD instance.
